### PR TITLE
Use caching secret manager in kubelet

### PR DESCRIPTION
**Pull request by wojtek-t**
I just found that this is in my local branch I'm using for testing, but not in master :)

### CODE REVIEW

**Comment by yujuhong** [comment]
/lgtm

Will keep an eye on the builds to see if anything abnormal pops up.

**Comment by smarterclayton** [comment]
Haha

**Comment by spxtr** [comment]
@k8s-bot kops aws e2e test this

**Comment by wojtek-t** [comment]
That is weird. It seems that kops is reliably failing, whereas gce is reliably passing :)

Anyway, those failure seem related to the PR, so I'm trying to understand them. So I started debugging it and logs from apiserver look pretty reasonable to me. For completeness pasting:
```
[test started]
...
// changes to secrets:
I0202 03:34:46.650781       5 wrap.go:75] DELETE /api/v1/namespaces/e2e-tests-secrets-skrrn/secrets/s-test-opt-del-897bc924-e8f8-11e6-a610-0242ac110002: (2.356989ms) 200 [[e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format] 146.148.94.48:37096]
I0202 03:34:46.681970       5 wrap.go:75] PUT /api/v1/namespaces/e2e-tests-secrets-skrrn/secrets/s-test-opt-upd-897bca2a-e8f8-11e6-a610-0242ac110002: (2.325301ms) 200 [[e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format] 146.148.94.48:37096]
...
// Now kubelet is syncing the pod (which triggered GETs to apiserver)
I0202 03:35:54.997743       5 wrap.go:75] GET /api/v1/namespaces/e2e-tests-secrets-skrrn/secrets/s-test-opt-upd-897bca2a-e8f8-11e6-a610-0242ac110002?resourceVersion=0: (365.139Âµs) 200 [[kubelet/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.55.95:34006]
I0202 03:35:54.998273       5 wrap.go:75] GET /api/v1/namespaces/e2e-tests-secrets-skrrn/secrets/s-test-opt-del-897bc924-e8f8-11e6-a610-0242ac110002?resourceVersion=0: (401.333Âµs) 404 [[kubelet/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.55.95:34006]
I0202 03:35:54.998776       5 wrap.go:75] GET /api/v1/namespaces/e2e-tests-secrets-skrrn/secrets/s-test-opt-create-897bca55-e8f8-11e6-a610-0242ac110002?resourceVersion=0: (361.381Âµs) 200 [[kubelet/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.55.95:34006]
```

And logs from kubelet from that time:
```
I0202 03:35:54.762737    2524 server.go:740] GET /containerLogs/e2e-tests-secrets-skrrn/pod-secrets-8985401f-e8f8-11e6-a610-0242ac110002/upds-volume-test: (1.809988ms) 200 [[Go-http-client/1.1] 172.20.35.149:58344]
I0202 03:35:54.981885    2524 operation_generator.go:514] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/898640c2-e8f8-11e6-9e11-0aa970757f80-deletes-volume" (spec.Name: "deletes-volume") pod "898640c2-e8f8-11e6-9e11-0aa970757f80" (UID: "898640c2-e8f8-11e6-9e11-0aa970757f80").
I0202 03:35:54.982090    2524 operation_generator.go:514] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/898640c2-e8f8-11e6-9e11-0aa970757f80-updates-volume" (spec.Name: "updates-volume") pod "898640c2-e8f8-11e6-9e11-0aa970757f80" (UID: "898640c2-e8f8-11e6-9e11-0aa970757f80").
I0202 03:35:54.982226    2524 operation_generator.go:514] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/898640c2-e8f8-11e6-9e11-0aa970757f80-creates-volume" (spec.Name: "creates-volume") pod "898640c2-e8f8-11e6-9e11-0aa970757f80" (UID: "898640c2-e8f8-11e6-9e11-0aa970757f80").
I0202 03:35:54.983025    2524 operation_generator.go:514] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/898640c2-e8f8-11e6-9e11-0aa970757f80-default-token-87v32" (spec.Name: "default-token-87v32") pod "898640c2-e8f8-11e6-9e11-0aa970757f80" (UID: "898640c2-e8f8-11e6-9e11-0aa970757f80").
I0202 03:35:56.795911    2524 server.go:740] GET /containerLogs/e2e-tests-secrets-skrrn/pod-secrets-8985401f-e8f8-11e6-a610-0242ac110002/upds-volume-test: (1.816318ms) 200 [[Go-http-client/1.1] 172.20.35.149:58344]
I0202 03:35:56.827797    2524 server.go:740] GET /containerLogs/e2e-tests-secrets-skrrn/pod-secrets-8985401f-e8f8-11e6-a610-0242ac110002/dels-volume-test: (1.575357ms) 200 [[Go-http-client/1.1] 172.20.35.149:58344]
```

So for now I don't understand this failure - everything looks as i would expect it.

That said, I added one unit test for checking of secret manager returns "not found" when secret is deletec, and temporily added some debug log to push debugging further.

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE

**Comment by wojtek-t** [comment]
I'm adding do-not-merge label until I debug it for now.

**Comment by wojtek-t** [comment]
But after I added more logs, I'm not longer able to reproduce it (neither here or locally). Trying more.

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE

**Comment by wojtek-t** [comment]
I think I have it. Basically I added the following logs to the code in secretStore.

```
secret, err := s.kubeClient.Core().Secrets(namespace).Get(name, opts)
// Update state, unless we got error different than "not-found".
glog.Errorf("SS: %s %s: %v %t", namespace, name, err, apierrors.IsNotFound(err))
if err == nil || apierrors.IsNotFound(err) {
	glog.Errorf("SS: %s %s: %v", secret)
```

And what I'm seeing in logs is this:

```
E0202 15:09:36.726287    1426 secret_manager.go:185] SS: e2e-tests-secrets-wnvw0 s-test-opt-del-6f8756ba-e959-11e6-a02d-0242ac110002: secrets "s-test-opt-del-6f8756ba-e959-11e6-a02d-0242ac110002" not found true
E0202 15:09:36.726304    1426 secret_manager.go:187] SS: &Secret{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Data:map[string][]byte{},Type:,StringData:map[string]string{},} %!s(MISSING): %!v(MISSING)
```

So it seems that apiserver is correctly returning a "not found error", but at the same time, the secret it returns is basically &v1.Secret{} (and not nil).


So with that, we have "resourceVersion" of the newSecret set to "" (which is interpreted as 0), which means it's older than the previous one. Which is bad.


I will fix it and add unit test for it tomorrow (or maybe I will try to find some time today evening).

**Comment by wojtek-t** [comment]
@yujuhong - I managed to reproduce the problem with unit test. The fix is already part of this PR (together with unit test for this).

PTAL

File: pkg/kubelet/secret/secret_manager.go [hunk 1]
```diff
@@ -181,7 +184,7 @@ func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
 		// Update state, unless we got error different than "not-found".
 		if err == nil || apierrors.IsNotFound(err) {
```
Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: I think this is slightly more readable then the logic you have. Feel free to disagree...
```go
if err != nil && !apierrors.IsNotFound(err) && (data.secret == nil && data.err == nil) {
    // Couldn't fetch the latest secret, but there is no cached data to return.
    // Return the fetch result instead.
    return secret, err
}

if (err == nil && !isSecretOlder(secret, data.secret)) || apierrorsIsNotFound(err){
    // If the fetch succeeded with a newer version of the secret, or if the secret could
    // not be found in the apiserver, update the cached data to reflect the current status.
    data.secret = secret
    data.err = err
    data.lastUpdateTime = s.clock.Now()
}

```

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: SGTM - changed

**Comment by wojtek-t** [comment]
@yujuhong - PTAL

**Pull request merged**